# Problem 2
Diberikan data mobile price classification sebagai berikut:   
https://www.kaggle.com/datasets/iabhishekofficial/mobile-price-classification  
Pada data tersebut terdapat beberapa fitur yang menggambarkan suatu produk mobile phone.  
Berdasarkan data tersebut:  
a. Hitunglah korelasi setiap fitur terhadap target klasifikasi, kemudian tuliskan top 5 fitur yang 
memiliki korelasi tertinggi terhadap target klasifikasi, berikan penjelasan anda.  
b. Buatlah suatu model klasifikasi dengan menggunakan algoritma yang anda ketahui, minimal 
anda menggunakan 2 algoritma yang dibandingkan performancenya. Lakukan beberapa tuning parameter untuk setiap algoritma kemudian analisa dan bandingkan performance  metric untuk setiap algoritma berdasarkan nilai accuracy, precision, recall dan F1 score.

**Solution for A**

1. Ambil data dari tabel, data yang saya akan ambil adalah data latih

In [93]:
import pandas as pd
 
data = pd.read_csv('train.csv')

2. Menghitung korelasi antara tiap fitur dengan target klasifikasi

In [94]:
correlation = data.corr()['price_range'].abs().sort_values(ascending=False)

3. Memperlihatkan top 5 fitur dengan korelasi tertinggi

In [95]:
top_5 = correlation[1:6] 

print("Top 5 fitur dengan korelasi tertinggi terhadap target klasifikasi:")
print(top_5)

Top 5 fitur dengan korelasi tertinggi terhadap target klasifikasi:
ram              0.917046
battery_power    0.200723
px_width         0.165818
px_height        0.148858
int_memory       0.044435
Name: price_range, dtype: float64


**Penjelasan**
- Fitur `ram` memiliki korelasi tertinggi dengan nilai $0.917046$. Hal ini menunjukkan bahwa nilai `ram` memiliki pengaruh yang kuat terhadap klasifikasi harga. Semakin tinggi nilai `ram`, kemungkinan besar harga juga akan lebih tinggi.
- Fitur `battery_power` memiliki korelasi sebesar $0.200723$, menunjukkan adanya hubungan yang sedikit positif antara `battery_power` dengan harga. Namun, korelasi ini tidak sekuat korelasi `ram`.
- Fitur `px_width` memiliki korelasi sebesar $0.165818$, menunjukkan adanya hubungan yang sedikit positif antara `px_width` dengan harga.
- Fitur `px_height` memiliki korelasi sebesar $0.148858$, menunjukkan adanya hubungan yang sedikit positif antara `px_height` dengan harga.
- Fitur `int_memory` memiliki korelasi sebesar $0.044435$, menunjukkan adanya hubungan yang sangat lemah antara `int_memory` dengan harga.

**Solution for B**

Model 1: Random Forest Classifier
1. Import library yang diperlukan, dan import data dari csv latih dan csv uji

In [96]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
 
train_data = pd.read_csv('train.csv') 

2. Pisahkan fitur dan target klasifikasi di data latih dan data uji. Disini saya tidak akan menggunakan csv uji karena tidak memiliki fitur `price_range`, karenanya saya akan melakukan split data dari data training.

In [97]:
X_train = train_data.drop('price_range', axis=1)
y_train = train_data['price_range']

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

3. Model pertama menggunakan Random Forest Classifier, lalu melakukan tuning parameter

In [98]:
rf_classifier = RandomForestClassifier() 

4. Melakukan pelatihan

In [99]:
rf_classifier.fit(X_train, y_train)

RandomForestClassifier()

5. Memilih model terbaik 

In [100]:
best_rf_model = rf_grid.best_estimator_

6. Membuat prediksi pada data test dengan menggunakan model Random Forest Classifier

In [101]:
rf_predictions = best_rf_model.predict(X_val)

Model 2: Gradient Boosting Clasifier

1. Membuat model 

In [102]:
gb_classifier = GradientBoostingClassifier()

2. Melakukan pelatihan

In [103]:

gb_classifier.fit(X_train, y_train)

GradientBoostingClassifier()

3. Mengambil model terbaik

In [104]:
best_gb_model = gb_grid.best_estimator_

5. Memprediksi target pada data test dengan model baru

In [105]:
gb_predictions = best_gb_model.predict(X_val)

Evaluasi:

In [106]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

rf_accuracy = accuracy_score(y_val, rf_predictions)
rf_precision = precision_score(y_val, rf_predictions, average='weighted')
rf_recall = recall_score(y_val, rf_predictions, average='weighted')
rf_f1_score = f1_score(y_val, rf_predictions, average='weighted')

gb_accuracy = accuracy_score(y_val, gb_predictions)
gb_precision = precision_score(y_val, gb_predictions, average='weighted')
gb_recall = recall_score(y_val, gb_predictions, average='weighted')
gb_f1_score = f1_score(y_val, gb_predictions, average='weighted')

# Menampilkan hasil performa
print("Random Forest Classifier:")
print("Accuracy:", rf_accuracy)
print("Precision:", rf_precision) 
print("Recall:", rf_recall)
print("F1 Score:", rf_f1_score)
print()

print("Gradient Boosting Classifier:")
print("Accuracy:", gb_accuracy)
print("Precision:", gb_precision)
print("Recall:", gb_recall)
print("F1 Score:", gb_f1_score)

Random Forest Classifier:
Accuracy: 0.895
Precision: 0.8970330611606846
Recall: 0.895
F1 Score: 0.8955327235564609

Gradient Boosting Classifier:
Accuracy: 0.92
Precision: 0.9207233445785458
Recall: 0.92
F1 Score: 0.9201407599753187


**Evaluasi**

Meskipun tanpa tuning parameter, keduanya menghasilkan tingkat akurasi yang tinggi, yaitu di atas $0.89$. Namun, Gradient Boosting Classifier menunjukkan performa yang sedikit lebih baik dengan nilai akurasi, precision, recall, dan F1 score yang lebih tinggi dibandingkan dengan Random Forest Classifier. Kesimpulannya, Gradient Boosting Classifier memberikan performa yang sedikit lebih unggul dan dinilai model yang baik dalam kasus iin.